kolibri/core/assets/src/conditionalPromise.js

```diff
@@ -0,0 +1,51 @@
+class ConditionalPromise {
+  /**
+   * Create a conditional promise - like a promise, but cancelable!
+   */
+  constructor(...args) {
+    if ([...args].length) {
+      this._promise = new Promise(...args);
+    }
+  }
+
+  catch(...args) {
+    this._promise.catch(...args);
+    return this;
+  }
+
+  then(...args) {
+    this._promise.then(...args);
+    return this;
+  }
+
+  only(cancelCheck, resolve, reject) {
```

Comment:
Cool pattern. Is this something you came up with, or is it modeled on something you saw?